From 117e3de0adda21e93e30907f3f93fbb7dff54993 Mon Sep 17 00:00:00 2001
From: Liu Shixin <liushixin2@huawei.com>
Date: Fri, 11 Nov 2022 15:13:29 +0800
Subject: [PATCH 0020/1398] [Huawei] mm/page_alloc: isolate
 percpu_max_batchsize with config SYSCTL_PERCPU_MAX_BATCHSIZE

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187965

--------------------------------

Isolate percpu_max_batchsize with config SYSCTL_PERCPU_MAX_BATCHSIZE.

Fixes: 97dc1be774d9 ("mm/page_alloc.c: add sysctl to revise the batch and high of percpu pageset")
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig | 1 +
 arch/x86/configs/openeuler_defconfig   | 1 +
 include/linux/mmzone.h                 | 4 +++-
 kernel/sysctl.c                        | 2 ++
 mm/Kconfig                             | 8 ++++++++
 mm/page_alloc.c                        | 2 ++
 6 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 13ef0294ab83..dc54e71f326a 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -1080,6 +1080,7 @@ CONFIG_ETMEM_SCAN=m
 CONFIG_ETMEM_SWAP=m
 CONFIG_ETMEM=y
 CONFIG_USERSWAP=y
+CONFIG_SYSCTL_PERCPU_MAX_BATCHSIZE=y
 CONFIG_CMA=y
 # CONFIG_CMA_DEBUG is not set
 # CONFIG_CMA_DEBUGFS is not set
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index 473635519202..9c6bf774bba0 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -1028,6 +1028,7 @@ CONFIG_ETMEM_SCAN=m
 CONFIG_ETMEM_SWAP=m
 CONFIG_ETMEM=y
 CONFIG_USERSWAP=y
+CONFIG_SYSCTL_PERCPU_MAX_BATCHSIZE=y
 # CONFIG_CMA is not set
 CONFIG_MEM_SOFT_DIRTY=y
 CONFIG_ZSWAP=y
diff --git a/include/linux/mmzone.h b/include/linux/mmzone.h
index fb313c795daf..7a1e442eccdc 100644
--- a/include/linux/mmzone.h
+++ b/include/linux/mmzone.h
@@ -1009,8 +1009,11 @@ int lowmem_reserve_ratio_sysctl_handler(struct ctl_table *, int, void *,
 		size_t *, loff_t *);
 int percpu_pagelist_fraction_sysctl_handler(struct ctl_table *, int,
 		void *, size_t *, loff_t *);
+#ifdef CONFIG_SYSCTL_PERCPU_MAX_BATCHSIZE
 int percpu_max_batchsize_sysctl_handler(struct ctl_table *, int,
 		void *, size_t *, loff_t *);
+extern int percpu_max_batchsize;
+#endif
 int sysctl_min_unmapped_ratio_sysctl_handler(struct ctl_table *, int,
 		void *, size_t *, loff_t *);
 int sysctl_min_slab_ratio_sysctl_handler(struct ctl_table *, int,
@@ -1018,7 +1021,6 @@ int sysctl_min_slab_ratio_sysctl_handler(struct ctl_table *, int,
 int numa_zonelist_order_handler(struct ctl_table *, int,
 		void *, size_t *, loff_t *);
 extern int percpu_pagelist_fraction;
-extern int percpu_max_batchsize;
 extern char numa_zonelist_order[];
 #define NUMA_ZONELIST_ORDER_LEN	16
 
diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 16bdc16a1e9f..db7f4c4a886f 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -3068,6 +3068,7 @@ static struct ctl_table vm_table[] = {
 		.proc_handler	= percpu_pagelist_fraction_sysctl_handler,
 		.extra1		= SYSCTL_ZERO,
 	},
+#ifdef CONFIG_SYSCTL_PERCPU_MAX_BATCHSIZE
 	{
 		.procname	= "percpu_max_batchsize",
 		.data		= &percpu_max_batchsize,
@@ -3076,6 +3077,7 @@ static struct ctl_table vm_table[] = {
 		.proc_handler	= percpu_max_batchsize_sysctl_handler,
 		.extra1		= SYSCTL_ZERO,
 	},
+#endif
 	{
 		.procname	= "page_lock_unfairness",
 		.data		= &sysctl_page_lock_unfairness,
diff --git a/mm/Kconfig b/mm/Kconfig
index e703e1d21dde..1ff990c1b894 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -566,6 +566,14 @@ config PAGE_CACHE_LIMIT
 
 	  if unsure, say N to disable the PAGE_CACHE_LIMIT.
 
+config SYSCTL_PERCPU_MAX_BATCHSIZE
+	bool "Support adjust percpu max batchsize directly"
+	depends on MMU && SYSCTL
+	default n
+	help
+	  Add a sysctl percpu_max_batchsize to adjust percpu batchsize and
+	  maxsize.
+
 config CMA
 	bool "Contiguous Memory Allocator"
 	depends on MMU
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 15fbf83846e0..01c18b8362b6 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -8676,6 +8676,7 @@ int percpu_pagelist_fraction_sysctl_handler(struct ctl_table *table, int write,
 	return ret;
 }
 
+#ifdef CONFIG_SYSCTL_PERCPU_MAX_BATCHSIZE
 int percpu_max_batchsize_sysctl_handler(struct ctl_table *table, int write,
 		void *buffer, size_t *length, loff_t *ppos)
 {
@@ -8708,6 +8709,7 @@ int percpu_max_batchsize_sysctl_handler(struct ctl_table *table, int write,
 	mutex_unlock(&pcp_batch_high_lock);
 	return ret;
 }
+#endif
 
 #ifndef __HAVE_ARCH_RESERVED_KERNEL_PAGES
 /*
-- 
2.17.1

